### PR TITLE
Optimize source refresh and ES indexation process

### LINKED ISSUE #111: Improve source refresh status in admin
Refresh status is get from async celery task.

These tasks are deleted in backend.

Sources are marked as "PENDING" when no data is find.

Solution:

- [x] Store results (success, error, pending, started) in Source.report .
  - state / start time / end time / result 
- [x] Read from this to provide info to admin.
- [x] Store all refresh infos in report (added / update / deleted count + exceptions)

### LINKED ISSUE #112: Improve source refresh process
Actually, a single transaction is performed to refresh a source.
This can take a very long time.

Improve refresh status by creating a new temporary layer, adding features, index to ES, THEN delete old one and linking new one. Perform database transaction per feature refreshed. If error occured, delete new layer and ES index.

### PR DESCRIPTION
**Pull request by LePetitTim**

### CODE REVIEW

**Review by CFarcy — DISMISSED** [review]

File: project/geosource/models.py [hunk 1]
```diff
@@ -63,6 +63,39 @@ def get_type_from_data(cls, data):
...
+    message = models.CharField(max_length=255, default="")
+    started = models.DateTimeField(null=True)
+    ended = models.DateTimeField(null=True)
+    added_lines = models.PositiveIntegerField(default=0)
+    deleted_lines = models.PositiveIntegerField(default=0)
+    modified_lines = models.PositiveIntegerField(default=0)
+    errors = models.JSONField(default=list)
+
+    def save(self, *args, **kwargs):
+        self.ended = timezone.now()
```
Comment (submarcos) [inline project/geosource/models.py:85]: I think that is task report method that should update this field

File: project/geosource/models.py [hunk 2]
```diff
@@ -76,7 +109,7 @@ class Source(PolymorphicModel, CeleryCallMethodsMixin):
 
     settings = models.JSONField(default=dict, blank=True)
     groups = models.ManyToManyField(Group, blank=True, related_name="geosources")
-    report = models.JSONField(default=dict, blank=True, encoder=DjangoJSONEncoder)
+    report = models.ForeignKey(SourceReporting, on_delete=models.PROTECT, null=True)
```
Comment (submarcos) [inline project/geosource/models.py:112]: report i nullable, so deleting report should set NULL on delete

Comment (CFarcy) [reply project/geosource/models.py:112]: Yes correct, I put protected when I start the dev and forgot to change it.

File: project/geosource/models.py [hunk 3]
```diff
@@ -63,6 +63,39 @@ def get_type_from_data(cls, data):
         return types.get(type(data), cls.Undefined)
 
 
+class SourceReporting(models.Model):
+    SUCCESS = "0"
+    ERROR = "1"
+    WARNING = "2"
```
[the remark this answers is not in the document]
Comment (CFarcy) [reply project/geosource/models.py:69]: Oh, nice !

File: project/geosource/models.py (hunk 2 above)
Comment (CFarcy) [reply project/geosource/models.py]: Done !

File: project/geosource/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CFarcy) [reply project/geosource/models.py]: Done !

File: project/geosource/models.py (hunk 1 above)
Comment (CFarcy) [reply project/geosource/models.py:82]: Are you talking about the methods in `project/geosource.task.py` ?

Comment (CFarcy) [reply project/geosource/models.py]: Done !

File: project/geosource/models.py [hunk 4]
```diff
@@ -63,7 +63,54 @@ def get_type_from_data(cls, data):
...
+    """Generic source exception to be catched by the generic Source Model"""
+
+    def __init__(self, message):
+        self.message = message
+
+
+class CSVSourceException(SourceException):
+    """CSVSource exception raised by the CSVSource model"""
+
+    pass
```
Comment (submarcos) [inline project/geosource/models.py:76]: can you move these in dedicated file ?

File: project/geosource/models.py [hunk 5]
```diff
@@ -76,13 +123,16 @@ class Source(PolymorphicModel, CeleryCallMethodsMixin):
...
-    report = models.JSONField(default=dict, blank=True, encoder=DjangoJSONEncoder)
+    report = models.OneToOneField(SourceReporting, on_delete=models.SET_NULL, null=True)
 
     task_id = models.CharField(null=True, max_length=255, blank=True)
     task_date = models.DateTimeField(null=True, blank=True)
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
     last_refresh = models.DateTimeField(default=timezone.now)
+    status = models.PositiveIntegerField(
+        choices=Status.choices, null=True, max_length=20
```
Comment (submarcos) [inline project/geosource/models.py:134]:
```suggestion
        choices=Status.choices, null=True
```
no need to get a max_length on an integer. Bu you can use PositiveSmallIntegerField

File: project/geosource/models.py [hunk 6]
```diff
@@ -63,7 +63,54 @@ def get_type_from_data(cls, data):
...
+
+    pass
+
+
+class SourceReporting(models.Model):
+    class Status(models.IntegerChoices):
+        SUCCESS = 0, "Success"
+        ERROR = 1, "Error"
+        WARNING = 2, "Warning"
+        PENDING = 3, "Pending"
```
[the remark this answers is not in the document]
Comment (CFarcy) [reply project/geosource/models.py:84]: Yes, you're right.  I think the report status (and even errors messages) could be translated values also !

File: project/geosource/models.py (hunk 4 above)
Comment (CFarcy) [reply project/geosource/models.py:76]: Yes, absolutely!

File: project/geosource/models.py (hunk 5 above)
Comment (CFarcy) [reply project/geosource/models.py:134]: Oh, that's nice. I will definitively use `PositiveSmallIntegerField`

**Pull request merged**